From stable-bounces@linux.kernel.org  Fri Jul 22 00:35:58 2005
Date: Fri, 22 Jul 2005 09:35:51 +0200
From: Patrick McHardy <kaber@trash.net>
User-Agent: Mozilla Thunderbird 1.0.2 (X11/20050602)
To: "David S. Miller" <davem@davemloft.net>
Cc: Netfilter Development Mailinglist <netfilter-devel@lists.netfilter.org>,
        stable@kernel.org
Subject: [PATCH][NETFILTER]: Wait until all references to ip_conntrack_untracked are dropped on unload

[NETFILTER]: Wait until all references to ip_conntrack_untracked are dropped on unload

Fixes a crash when unloading ip_conntrack.

Signed-off-by: Patrick McHardy <kaber@trash.net>
Signed-off-by: Chris Wright <chrisw@osdl.org>
---

 net/ipv4/netfilter/ip_conntrack_core.c |    3 +++
 1 files changed, 3 insertions(+)

Index: linux-2.6.12.y/net/ipv4/netfilter/ip_conntrack_core.c
===================================================================
--- linux-2.6.12.y.orig/net/ipv4/netfilter/ip_conntrack_core.c
+++ linux-2.6.12.y/net/ipv4/netfilter/ip_conntrack_core.c
@@ -1124,6 +1124,9 @@ void ip_conntrack_cleanup(void)
 		schedule();
 		goto i_see_dead_people;
 	}
+	/* wait until all references to ip_conntrack_untracked are dropped */
+	while (atomic_read(&ip_conntrack_untracked.ct_general.use) > 1)
+		schedule();
 
 	kmem_cache_destroy(ip_conntrack_cachep);
 	kmem_cache_destroy(ip_conntrack_expect_cachep);
